### PR TITLE
File Output Setting

### PR DESCRIPTION
**Pull request by calgray**
This adds support for specifying a file output rather than printing to the terminal either via CLI or config settings. Currently this only ovewrites the file at the specified path and reports an error early if the specified file cannot be written to.

I've also started adding more to the project documentation too make it easier to understand the project layout and terminology.

### CODE REVIEW

File: src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu [hunk 1]
```diff
@@ -95,6 +95,13 @@ namespace cuda
 
         // Flooring to remove incomplete measurements
         int integrations = ms.GetNumRows() / ms.GetNumBaselines();
+        if(integrations == 0)
+        {
+            std::stringstream ss;
+            ss << "invalid number of rows, expected >" << ms.GetNumBaselines() << ", got " << ms.GetNumRows();
+            throw icrar::file_exception(ms.GetFilepath().get_value_or("unknown"), ss.str(), __FILE__, __LINE__);
+        }
```
Comment (calgray) [inline src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu:103]: Without this there was a "cannot iterate scalears" error on some measurement sets that caused confusion

File: src/icrar/leap-accelerate-cli/Arguments.h [hunk 2]
```diff
@@ -114,6 +114,9 @@ namespace icrar
          */
         std::unique_ptr<MeasurementSet> m_measurementSet;
         std::istream* m_inputStream = nullptr; // Cached reference to the input stream
+        
+        std::unique_ptr<std::ofstream> m_outputFileStream;
+        std::ostream* m_outputStream = nullptr;
```
Comment (rtobar) [inline src/icrar/leap-accelerate-cli/Arguments.h:119]: `m_outputStream` doesn't need to be stored separately, it's already contained in the `unique_ptr`

File: src/icrar/leap-accelerate-cli/Arguments.cc [hunk 3]
```diff
@@ -126,6 +126,23 @@ namespace icrar
...
+        {
+            m_outputFileStream = std::make_unique<std::ofstream>(m_outputFilePath.get() + ".json");
+            if(!m_outputFileStream->is_open())
+            {
+                std::stringstream ss;
+                ss << "failed to open file " << m_outputFilePath.get() << ".json" << std::endl;
+                throw exception(ss.str(), __FILE__, __LINE__);
+            }
+
+            m_outputStream = m_outputFileStream.get();
```
Comment (rtobar) [inline src/icrar/leap-accelerate-cli/Arguments.cc:140]: Not needed

File: src/icrar/leap-accelerate-cli/Arguments.cc [hunk 4]
```diff
@@ -196,6 +213,16 @@ namespace icrar
...
     }
 
+    boost::optional<std::string> ArgumentsValidated::GetOutputFilePath() const
+    {
+        return m_outputFilePath;
+    }
+
+    std::ostream& ArgumentsValidated::GetOutputStream()
+    {
+        return *m_outputStream;
```
Comment (rtobar) [inline src/icrar/leap-accelerate-cli/Arguments.cc:223]: `return *m_outputFileStream.get()`

**Review by rtobar — COMMENTED** [review]
Looks good, except that I think there's no need to store `m_outputStream` given that you are already storing `m_outputFileStream`, which is what my comments are about.

File: src/icrar/leap-accelerate-cli/Arguments.cc (hunk 4 above)
Comment (calgray) [reply src/icrar/leap-accelerate-cli/Arguments.cc:223]: I store m_outputStream seperately since in some cases m_outputStream is set to std::cout

Comment (rtobar) [reply src/icrar/leap-accelerate-cli/Arguments.cc:223]: Ahhhh right, my bad, I didn't see that the first time. All good then :)

**Review by rtobar — APPROVED** [review]

File: src/icrar/leap-accelerate-cli/Arguments.cc (hunk 4 above)
Comment (rtobar) [reply src/icrar/leap-accelerate-cli/Arguments.cc:223]: Thinking about it again, an alternative would be to only store `m_outputFileStream`, then implement `GetOutputStream()` like:

```c++
std::ostream &ArgumentsValidated::GetOutputStream()
{
    if (m_outputFileStream)
    {
        return *m_outputFileStream.get();
    }
    return std::cerr;
}
```

Comment (calgray) [reply src/icrar/leap-accelerate-cli/Arguments.cc:223]: Pretty much, the only difference between the two is m_outputStream is acting as a cache variable

**Pull request merged**